packages/sitecore-jss-proxy/src/personalize/PersonalizeHelper.ts

```diff
@@ -0,0 +1,310 @@
+              params,
+              language,
+              timeout,
+            },
+            req
+          ).then((personalization) => {
+            const variantId = personalization.variantId;
+            if (variantId) {
+              if (!execution.variantIds.includes(variantId)) {
+                debug.personalize('invalid variant %s', variantId);
+              } else {
+                identifiedVariantIds.push(variantId);
+              }
+            }
+          })
+        )
+      );
+    } catch (e) {
+      debug.personalize('skipped, error %o', e);
+    }
+
+    if (identifiedVariantIds.length === 0) {
+      debug.personalize('skipped (no variant(s) identified)');
+      return [];
+    }
+    return identifiedVariantIds;
+  };
+
+  protected personalizeLayout(layoutData: LayoutServiceData, variantIds: string[]) {
+    if (!layoutData.sitecore?.route) {
+      debug.personalize('skipped (layout is empty)');
+      return layoutData;
+    }
+    const personalizeData = getGroomedVariantIds(variantIds);
+    const personalizedPlaceholders = personalizeLayout(
+      layoutData,
+      personalizeData.variantId,
+      personalizeData.componentVariantIds
+    );
+    layoutData.sitecore.route.placeholders = personalizedPlaceholders;
+    return layoutData;
+  }
+
+  protected getLanguage(layoutData: LayoutServiceData): string {
+    return layoutData.sitecore?.context?.language || 'en';
+  }
+
+  protected getHostHeader(req: IncomingMessage): string {
+    return req.headers.host?.split(':')[0] || '';
+  }
+
+  protected excludeRoute(pathname: string) {
+    return this.config?.excludeRoute && this.config?.excludeRoute(pathname);
+  }
+
+  protected extractDebugHeaders(incomingHeaders: IncomingHttpHeaders) {
+    const headers = {} as { [key: string]: string | string[] | undefined };
+    Object.keys(incomingHeaders).forEach(
+      (key) => incomingHeaders[key] && (headers[key] = incomingHeaders[key])
+    );
+    return headers;
+  }
+
+  protected async personalize(
+    {
+      params,
+      friendlyId,
+      language,
+      timeout,
+      variantIds,
+    }: {
+      params: ExperienceParams;
+      friendlyId: string;
+      language: string;
+      timeout?: number;
+      variantIds?: string[];
+    },
+    request: IncomingMessage
+  ) {
+    debug.personalize('executing experience for %s %o', friendlyId, params);
+
+    return (await personalize(
+      request,
+      {
+        channel: this.config.cdpConfig.channel || 'WEB',
+        currency: this.config.cdpConfig.currency ?? 'USD',
+        friendlyId,
+        params,
+        language,
+        pageVariantIds: variantIds,
+      },
+      { timeout }
+    )) as {
+      variantId: string;
+    };
+  }
+
+  protected getExperienceParams(req: IncomingMessage): ExperienceParams {
+    // nodejs req.url does not have a hostname, we parse query string the old fashioned way
+    const rawQs = req.url?.split('?')[1] || '';
```

Comment:
Suggest you to use more "safe" approach
```ts
const rawQs = req.url ? new URL(req.url).searchParams : '';
```

Reply:
req.url does not have a hostname - we cannot transform it with `new URL` reliably.

Reply:
got it
makes sense now